### DISCUSSION TITLE
.timezone() with zone id doesn't adjust the date but only the time

### DISCUSSION THREAD

**Discussion by amanseth123** [shred/commons-suncalc#43] [Q&A]
I'm working on an application where I need to calculate sunrise sunset time of given location with latitude/longitude, local date in joda.time, twilight, and GMT/UTC with offset values. I'm getting incorrect result where the sunset appears to happen before sunrise on the same day despite of any offset value. I've attached a piece of code below which I'm using 
`SunTimes sunTimes = SunTimes.compute().on(localDate.getYear(), localDate.getMonthOfYear(), localDate.getDayOfMonth())
        .timezone(ZoneId.of("GMT-9"))
        .at(latitude,longitude)
        .twilight(zenith)
        .execute();`
Returned Output: SunTimes[rise=2010-07-02T23:10:37-09:00[GMT-09:00], set=2010-07-02T18:41:41-09:00[GMT-09:00], noon=2010-07-02T08:55:43-09:00[GMT-09:00], nadir=2010-07-02T20:55:54-09:00[GMT-09:00], alwaysUp=false, alwaysDown=false]
Here we can clearly see that the sunset is happening before the sunrise on the same day which is not correct.
Suggest if something I need to take care of while working with such scenario.

**Comment by shred** [comment]
You usually get this kind of result if the timezone does not match the given coordinates (e.g. location is Paris, timezone is New York).

Can you give me the a location, so I can reproduce the output?

**Reply by amanseth123** [reply]
I want to calculate sunrise/sunset time for ORD airport which is in Chicago(41.9765,-87.9051), and uses GMT-6(I'm using this as the time zone instead of using America/Chicago) timezone. And my system default timezone suppose is UTC+5:30. 
so I'm taking timezone id as GMT-6 and the local date as a DateTime object from joda.time.
Also `SunTimes sunTimes = SunTimes.compute().on(localDate.getYear(), localDate.getMonthOfYear(), localDate.getDayOfMonth()) .timezone(ZoneId.of("GMT-9")) .at(latitude,longitude) .twilight(zenith) .execute();` considering the code, the location coordinates corresponds to the the input timezone to .timezone() then how still it is returning wrong? And suppose my code server is in Asia somewhere picking up the system default time zone of Asia, but I'm trying to calculate sunrise sunset time for ORD(Chicago) in UTC so according to you the result will have such discrepancies? Will be glad to have your inputs to this
@shred

**Reply by shred** [reply]
```java
SunTimes sunTimes = SunTimes.compute()
                .on(2010, 7, 2)
                .timezone(ZoneId.of("America/Chicago"))
                .at(41.978603, -87.904842)
                .execute();
```
gives:
```
SunTimes[rise=2010-07-02T05:20:35-05:00[America/Chicago], set=2010-07-02T20:30:57-05:00[America/Chicago], noon=2010-07-02T12:55:43-05:00[America/Chicago], nadir=2010-07-02T00:55:42-05:00[America/Chicago], alwaysUp=false, alwaysDown=false]
```

Which is the correct time of sunrise and sunset at ORD airport on July 2nd 2010.

The correct time zone for ORD is CST, which is `UTC-6` (`UTC-5` on daylight saving time).

What you are asking is: "When does the sun rise and set in ORD, starting at July 2nd 2010, midnight, GMT-9?". Related to that position, the sun can actually set before it raises in ORD.

The other time difference probably comes from `.twilight(zenith)`, which contains an unknown value to me.

**Comment by shred** [comment]
GMT-9 is more or less the daylight savings time at Hawaii. What you are asking in your code is: "I am at Hawaii. Starting July 2nd 2010, midnight **Hawaiian time**, when is the next sunrise and sunset in Chicago?" The answer is correct that you will get a sunset first, and then a sunrise.

What you need to do is to set the start time of the computation to midnight **Chicago time**. For example:

```java
        SunTimes sunTimes = SunTimes.compute()
                .on(ZonedDateTime.of(2010, 7, 2, 0, 0, 0, 0, ZoneId.of("America/Chicago")))
                .timezone(ZoneId.of("GMT-9"))
                .at(41.978603, -87.904842)
                .execute();
```

This will give the next sunrise and sunset in Chicago, but the result is at "GMT-9" timezone. I guess this is what you want?

**Reply by amanseth123** [reply]
Well I'm using a local date and not a zonedDateTime which is why maybe I'm getting the result that way. Let me try the way you just suggested.
I'm doing something like this to get the sunrise time 
`SunTimes sunTimes = SunTimes.compute().on(date.getYear(), date.getMonthOfYear(), date.getDayOfMonth())
                .timezone(UTC + offset)
                .at(latitude, longitude)
                .twilight(twilight)
                .execute();
        ZonedDateTime sunriseTimeZoned = sunTimes.getRise();
        DateTime sunriseTimeUTC = new DateTime(sunriseTimeZoned.toInstant().toEpochMilli(), DateTimeZone.forOffsetHours(offset.getTotalSeconds() / 3600));
        return sunriseTimeUTC;`
Correct me if I'm doing anything wrong here

**Reply by shred** [reply]
To be honest, it is still a bit unclear to me what result you actually expect. If you need the `sunriseTimeUTC` as UTC time, you can do `SunTimes.compute().utc(). ...` (instead of `.timezone()`, and get the _result_ in UTC time zone.

You would still need to set the correct starting hour for your target location at `.on()`, like in my example above.
